doc/design/sql_program_dependency_analyze.md

```diff
@@ -0,0 +1,161 @@
+start only when statement 1 is finished.
+
+We can do this by following below steps:
+
+1. Parse every SQL statement, return a map recording read/write type of the tables that the current
+   statement is manipulating. We'll get a list of this map in order, representing each SQL statement.
+   This will be done in the third-party parsers.
+2. Parse the SQLFlow extended SQL if there are any, return a similar map recording read/write type
+   of the models of the current statement (in our Go parser).
+3. Construct a graph using the following rules:
+    1. The statements are nodes in the graph, they should run in the original order.
+    2. If one statement reads a table/model, then construct one graph edge from statements that write
+       the table to the current statement.
+4. Execute this graph by submitting the graph as an Argo/Tekton workflow, each step is one single
+   statement.
+
+## Analyzing Database Context
+
+Some SQL program have `USE` clause which defines the database the following SQL statements are using.
+
+```SQL
+USE db;
+CREATE TABLE my_table AS ...
+SELECT * FROM my_table TO TRAIN ...
+USE db2;
+SELECT * FROM my_table_in_db2 TO TRAIN ...
+```
+
+We need to go over the SQL program and get the last `USE` for each SQL statement, and modify the table
+name with out prefix `db.` to `db.table` for later dependency analyzing.
+
+## Hazard
+
+SQL program is just like normal computer programs. We treat tables as variables in our case.
+Analyzing computer programs always have hazards: https://en.wikipedia.org/wiki/Hazard_(computer_architecture). In the example above, we can construct a graph like below:
+
+<p align="center">
+<img src="figures/hazard.png">
+</p>
+
+
+Note that the last "DROP TABLE" statement must execute after the two "Explain" statements, because
+the table is used by the explain statements before it could be changed (in this case, deleted). This
+case is called "Write After Read". To solve this kind of data hazards, we can add one dependency node
+"table WAR" between the explain statements and drop statement:
+
+<p align="center">
+<img src="figures/hazard_solve.png">
+</p>
+
+We also need to solve "Write After Write" hazard just like "Write After Read" if the SQL program has
+this situation.
+
+## Implementation
+
+In order to construct the dependency graph of the SQL program, we need to analyze the parsed SQL
+statements in every parser, including MySQL parser, Hive parser, calcite parser and SQLFlow extended
+parser. Since the Hive parser and calcite parser is written in Java, we need to pass the table read/write
+information from Java to Go:
+
+```proto
+message ParserResponse {
+  repeated string sql_statements = 1;
+  int32 index = 2;
+  string error = 3;
+  // return tables that a statement manipulate.
+  repeated string input_tables = 4;
+  repeated string output_tables = 5;
+}
+```
+
+When parsing the SQL program in Go, we can then get table read/write information for each standard
```

Comment:
It seems that here we need a section describing how to change the SQLFlow syntax extension parser to retrieve the input and output models.